### ISSUE TITLE
Bnd's Maven dependency scope handling not correct

### ISSUE THREAD

**Issue by maggu2810** [bndtools/bnd#3783]
For me it seems that the logic bnd uses to handle resolve the Maven dependency scope is not correct.

To demonstrate the mismatch I add an example here: https://github.com/maggu2810/bnd-depscope.tmp

There are to POM file that define dependencies:
* deps/compile: dependencies used at compile time
* deps/runtime: dependencies used at runtime

There is one application that depends on both artifact:
* deps compile using scope provided
* deps runtime using scope compile

So, if the Maven dependency scope are applied correctly the dependencies of "deps compile" that are not part of "deps runtime" should use scope provided.
The Maven dependencies that are part of "deps runtime" should use scope "compile".

Maven handles this correctly.

After checking out the repo above execute `mvn clean install`.
This will build the whole reactor.

Now, inspect the dependencies and their copy of the "app.demo": `mvn dependency:list -pl apps/demo/`

It identifies that the json-b and json specification is using "compile" scope.
* org.apache.geronimo.specs:geronimo-jsonb_1.0_spec:jar:1.1:compile
* org.apache.geronimo.specs:geronimo-json_1.1_spec:jar:1.2:compile

If you open the `demo.bndrun` file in the Eclipse IDE you will realize that these bundles are not part of the implicit repository (but it should as compile and runtime scope should be added).

I assume artifacts that are added twice, the first time using scope provided and the second time using scope compile are not correctly resolved by Bnd. The "new" scope will be compile (as done by Maven itself) but Bnd stays at provided.

**Comment by maggu2810** [comment]
@rotty3000 You already fixed #3361 that has been about a different dependency handling of maven compiler plugin vs. bnd's maven plugins.

Do you think you will some time to look into this one?

**Comment by maggu2810** [comment]
Hi,
I started some debug session and it seems I found the cause of the problem.

Here a short example that should demonstrate the problem.

Think about a POM file that looks like this one:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <parent>
    <groupId>de.maggu2810.bnd.depscope.deps</groupId>
    <artifactId>de.maggu2810.bnd.depscope.deps</artifactId>
    <version>1.0.0-SNAPSHOT</version>
  </parent>

  <artifactId>de.maggu2810.bnd.depscope.deps.compile</artifactId>
  <packaging>pom</packaging>

  <dependencies>
    <dependency>
      <groupId>org.apache.geronimo.specs</groupId>
      <artifactId>geronimo-json_1.1_spec</artifactId>
      <version>1.2</version>
    </dependency>
  </dependencies>

</project>
```

And another one that looks like this one:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <parent>
    <groupId>de.maggu2810.bnd.depscope.deps</groupId>
    <artifactId>de.maggu2810.bnd.depscope.deps</artifactId>
    <version>1.0.0-SNAPSHOT</version>
  </parent>

  <artifactId>de.maggu2810.bnd.depscope.deps.runtime</artifactId>
  <packaging>pom</packaging>

  <dependencies>
    <dependency>
      <groupId>org.apache.geronimo.specs</groupId>
      <artifactId>geronimo-jsonb_1.0_spec</artifactId>
      <version>1.1</version>
    </dependency>
  </dependencies>

</project>
```

Nothin special.

The "deps.comile" depends on the JSON-P API.
The "deps.runtime" depends on the JSON-B API.

Now let's create a simple bnd application ("app.demo") that contains that dependencies:

```
  <dependencies>
    <dependency>
      <groupId>de.maggu2810.bnd.depscope.deps</groupId>
      <artifactId>de.maggu2810.bnd.depscope.deps.compile</artifactId>
      <version>1.0.0-SNAPSHOT</version>
      <type>pom</type>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>de.maggu2810.bnd.depscope.deps</groupId>
      <artifactId>de.maggu2810.bnd.depscope.deps.runtime</artifactId>
      <version>1.0.0-SNAPSHOT</version>
      <type>pom</type>
      <scope>runtime</scope>
    </dependency>
  </dependencies>
```

The "deps.compile" is a dependency using scope provided.
The "deps.runtime" is a dependency using scope runtime.

Let's look at the dependency tree:

```
[INFO] de.maggu2810.bnd.depscope.apps:de.maggu2810.bnd.depscope.app.demo:jar:1.0.0-SNAPSHOT
[INFO] +- de.maggu2810.bnd.depscope.deps:de.maggu2810.bnd.depscope.deps.compile:pom:1.0.0-SNAPSHOT:provided
[INFO] |  \- org.apache.geronimo.specs:geronimo-json_1.1_spec:jar:1.2:provided
[INFO] \- de.maggu2810.bnd.depscope.deps:de.maggu2810.bnd.depscope.deps.runtime:pom:1.0.0-SNAPSHOT:runtime
[INFO]    \- org.apache.geronimo.specs:geronimo-jsonb_1.0_spec:jar:1.1:runtime
```

* deps.compile uses scope provided
* JSON-P API uses scope provided
* deps.runtime uses scope runtime
* JSON-B API uses scope runtime

All fine.

Next step...

Now we add the JSON-P API additionally(!) to the "deps.runtime" POM.
So JSON-P API is part of "deps.compile" AND "deps.runtime".

Let'S look at the dependency tree:

```
[INFO] de.maggu2810.bnd.depscope.apps:de.maggu2810.bnd.depscope.app.demo:jar:1.0.0-SNAPSHOT
[INFO] +- de.maggu2810.bnd.depscope.deps:de.maggu2810.bnd.depscope.deps.compile:pom:1.0.0-SNAPSHOT:provided
[INFO] |  \- org.apache.geronimo.specs:geronimo-json_1.1_spec:jar:1.2:runtime
[INFO] \- de.maggu2810.bnd.depscope.deps:de.maggu2810.bnd.depscope.deps.runtime:pom:1.0.0-SNAPSHOT:runtime
[INFO]    \- org.apache.geronimo.specs:geronimo-jsonb_1.0_spec:jar:1.1:runtime
```

Please identify:
* Every artifact is listet only once.
* The scope of the JSON-P API changed from provided to runtime
* As every artifat is listet only once, it is still a child of deps.compile
* deps.compile is still using scope provided



To build the implicit repository the class [DependencyResolver](https://github.com/bndtools/bnd/blob/f95916e57965b85d9b95fba1743456dba8ba9db4/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/DependencyResolver.java) is used.
* `MavenImplicitProjectReposioty.createRepo` line 131 calls `BndrunContainer.getFileSetRepository`
* `BndrunContainer.getFileSetRepository` line 249 calls `DependencyResolver.getFileSetRepository`
* `DependencyResolver.getFileSetRepository` line 152 calls `DependencyResolver.resolve`
* ...

* a dependency resolution result is created: https://github.com/bndtools/bnd/blob/f95916e57965b85d9b95fba1743456dba8ba9db4/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/DependencyResolver.java#L127-L132
* a dependency graph is created to that result: https://github.com/bndtools/bnd/blob/f95916e57965b85d9b95fba1743456dba8ba9db4/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/DependencyResolver.java#L136
* the artifacts are discovered for that dependency graph: https://github.com/bndtools/bnd/blob/f95916e57965b85d9b95fba1743456dba8ba9db4/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/DependencyResolver.java#L140-L141

The artifacts discovery [iterates over the dependency nodes](https://github.com/bndtools/bnd/blob/f95916e57965b85d9b95fba1743456dba8ba9db4/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/DependencyResolver.java#L176).
If the scope of the dependency [is not part of the given scope that whole node is skipped](https://github.com/bndtools/bnd/blob/f95916e57965b85d9b95fba1743456dba8ba9db4/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/DependencyResolver.java#L177-L180) -- and so every artifact that is part of that node

This seems to be wrong to me.
Remember: Every artifact is only listet once in the graph.
So, you could get an "runtime" scope below an "provided" scope (look at the tree above).

The code currently skips "deps.compile" and all artifacts of that scope and so will skip the JSON-P API regardless that Maven states it will be part of the runtime scope artifacts.

IMHO the scope should be used to decide if an artifact is added to the files map that contains the result.
But it should not be used to skip a whole tree with all its children that are allowed to use another scope.

WDYT?

**Comment by maggu2810** [comment]
Suggested change (will create a PR if you agree):

```
diff --git a/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/DependencyResolver.java b/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/DependencyResolver.java
index 88fc207b8..526155a2f 100644
--- a/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/DependencyResolver.java
+++ b/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/DependencyResolver.java
@@ -174,11 +174,6 @@ public class DependencyResolver {
                List<RemoteRepository> remoteRepositories) throws MojoExecutionException {
 
                for (DependencyNode node : nodes) {
-                       if (!scopes.contains(node.getDependency()
-                               .getScope())) {
-                               continue;
-                       }
-
                        List<RemoteRepository> combinedRepositories = new ArrayList<>(remoteRepositories);
                        combinedRepositories.addAll(node.getRepositories());
 
@@ -189,12 +184,17 @@ public class DependencyResolver {
                                logger.debug("Located file: {} for artifact {}", resolvedArtifact.getArtifact()
                                        .getFile(), resolvedArtifact);
 
-                               files.put(resolvedArtifact.getArtifact()
-                                       .getFile(), resolvedArtifact);
+                               // Add artifact only if the scope of this artifact matches.
+                               if (scopes.contains(node.getDependency()
+                                       .getScope())) {
+                                       files.put(resolvedArtifact.getArtifact()
+                                               .getFile(), resolvedArtifact);
+                               }
                        } catch (ArtifactResolutionException e) {
                                throw new MojoExecutionException("Failed to resolve the dependency " + node.getArtifact()
                                        .toString(), e);
                        }
+
                        if (includeTransitive) {
                                discoverArtifacts(files, node.getChildren(), node.getRequestContext(), combinedRepositories);
                        } else {
```

**Comment by rotty3000** [comment]
@maggu2810 this is great investigation. I understand the flaw about cutting off children when a dependency's scope is not in the list. I'm fine with solving that goal.

However can you point me at documentation for using values other than `<scope>import</scope>` of the BOM dependency? I want to understand the mechanics involved in that.

**Comment by maggu2810** [comment]
WRT http://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope

The scope _import_ is supported for dependencies of type pom in `<dependencyManagement>` only. It is used for the BOM pattern.
So, you can add a GA to your dependencies and the other options (mainly V(ersion)) is taken from the BOM.

But that's not used in my example at all.
The pom files are added as dependencies (not management) to add all its transitive dependencies to the artifact.

If this makes sense, can be done be better, is completely wrong or if there is some documentation... I don't know

**Comment by rotty3000** [comment]
sorry this should have been closed by merge of #3880
